apps/ejabberd/src/ejabberd_cowboy.erl

```diff
@@ -121,22 +124,13 @@ do_start_cowboy(Ref, Opts, Retries, SleepTime) ->
     end.
 
 do_start_cowboy(Ref, Opts) ->
-    Port = gen_mod:get_opt(port, Opts),
-    IP = gen_mod:get_opt(ip, Opts, {0, 0, 0, 0}),
     SSLOpts = gen_mod:get_opt(ssl, Opts, undefined),
     NumAcceptors = gen_mod:get_opt(num_acceptors, Opts, 100),
-    MaxConns = gen_mod:get_opt(max_connections, Opts, 1024),
-    Compress = gen_mod:get_opt(compress, Opts, false),
-    Middlewares = case gen_mod:get_opt(middlewares, Opts, undefined) of
-        undefined -> [];
-        M -> [{middlewares, M}]
-    end,
-    TransportOpts = [{port, Port},
-                     {ip, IP},
-                     {max_connections, MaxConns}
-                    ],
-    Dispatch = cowboy_router:compile(get_routes(gen_mod:get_opt(modules, Opts))),
-    ProtocolOpts = [{compress, Compress}, {env, [{dispatch, Dispatch}]} | Middlewares],
+    TransportOpts = gen_mod:get_opt(transport_options, Opts, []),
+    Modules = gen_mod:get_opt(modules, Opts),
+    Dispatch = cowboy_router:compile(get_routes(Modules)),
+    ProtocolOpts = lists:merge([{env, [{dispatch, Dispatch}]}],
```

Comment:
The `merge` is for merging two independent sorted lists into one sorted list. I believe its behaviour is undefined for unsorted list (but the implementation happens to work). This use case simply calls for `[{env, [{dispatch, Dispatch}]} | gen_mod:get_opt(protocol_options, Opts, [])]`